### PR TITLE
Cmdlets for Azure RM SQL Long Term Retention Backup/Restore

### PR DESCRIPTION
**Pull request by hchungmsft**
- [x] publish new azure-sdk-for-net nuget package 0.47(http://azuresdkci.cloudapp.net/view/2-.NET%20SDK/job/sdk-for-net-publish/)
- [x] add new helpfile

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@hchungmsft__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by hovsepm** [comment]
@hchungmsft please fix build breaks

File: ChangeLog.md [hunk 1]
```diff
@@ -1,3 +1,12 @@
+##2016.07.04 version 1.6.0
+* Azure SQL DB Backup/Restore
+  * Cmdlets for LTR (Long Term Retention) backup feature
+	* Get-AzureRmSqlServerBackupArchivalVault
```
Comment (hchungmsft) [inline ChangeLog.md]: should be LongTermRetention

File: .gitignore [hunk 2]
```diff
@@ -199,5 +199,6 @@ FakesAssemblies/
 
 .vs/
 /tools/*.dll
+/tools/LocalFeed
```
Comment (markcowl) [inline .gitignore]: Remove this.  LocalFeed is sometimes used, we should not ignore its contents (but they shoudl be empty except under very special circumstances

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/AzureSqlDatabaseBackupLongTermRetentionPolicyCmdletBase.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+using Microsoft.Azure.Commands.Common.Authentication.Models;
+using Microsoft.Azure.Commands.Sql.Backup.Model;
+using Microsoft.Azure.Commands.Sql.Backup.Services;
+using Microsoft.Azure.Commands.Sql.Common;
+using Microsoft.Azure.Commands.Sql.Database.Model;
+using Microsoft.Azure.Commands.Sql.Database.Services;
+
+namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
+{
+    public abstract class AzureSqlDatabaseBackupLongTermRetentionPolicyCmdletBase : AzureSqlCmdletBase<IEnumerable<AzureSqlDatabaseBackupLongTermRetentionPolicyModel>, AzureSqlDatabaseBackupAdapter>
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/AzureSqlDatabaseBackupLongTermRetentionPolicyCmdletBase.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/AzureSqlServerBackupLongTermRetentionVaultCmdletBase.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+using Microsoft.Azure.Commands.Common.Authentication.Models;
+using Microsoft.Azure.Commands.Sql.Backup.Model;
+using Microsoft.Azure.Commands.Sql.Backup.Services;
+using Microsoft.Azure.Commands.Sql.Common;
+using Microsoft.Azure.Commands.Sql.Database.Model;
+using Microsoft.Azure.Commands.Sql.Database.Services;
+
+namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
+{
+    public abstract class AzureSqlServerBackupLongTermRetentionVaultCmdletBase : AzureSqlCmdletBase<IEnumerable<AzureSqlServerBackupLongTermRetentionVaultModel>, AzureSqlDatabaseBackupAdapter>
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/AzureSqlServerBackupLongTermRetentionVaultCmdletBase.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureRMSqlDatabaseGeoBackup.cs [hunk 5]
```diff
@@ -18,7 +18,7 @@
 
 namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
 {
-    [Cmdlet(VerbsCommon.Get, "AzureRMSqlDatabaseGeoBackup",
+    [Cmdlet(VerbsCommon.Get, "AzureRmSqlDatabaseGeoBackup",
         ConfirmImpact = ConfirmImpact.None)]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureRMSqlDatabaseGeoBackup.cs]: Never up[date ConfirmImpact without using SupportsShouldProcess=true

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureRMSqlDeletedDatabaseBackup.cs [hunk 6]
```diff
@@ -19,7 +19,7 @@
 
 namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
 {
-    [Cmdlet(VerbsCommon.Get, "AzureRMSqlDeletedDatabaseBackup",
+    [Cmdlet(VerbsCommon.Get, "AzureRmSqlDeletedDatabaseBackup",
         ConfirmImpact = ConfirmImpact.None)]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureRMSqlDeletedDatabaseBackup.cs]: same comment

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+using System.Linq;
+using System.Collections.Generic;
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Sql.Backup.Model;
+using Microsoft.Azure.Commands.Sql.Database.Model;
+
+namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
+{
+    [Cmdlet(VerbsCommon.Get, "AzureRmSqlDatabaseBackupLongTermRetentionPolicy",
+        ConfirmImpact = ConfirmImpact.None)]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs]: same comment

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        /// <summary>
+        /// Get the entities from the service
+        /// </summary>
+        /// <returns>The list of entities</returns>
+        protected override IEnumerable<AzureSqlDatabaseBackupLongTermRetentionPolicyModel> GetEntity()
+        {
+            ICollection<AzureSqlDatabaseBackupLongTermRetentionPolicyModel> results;
+
+            results = new List<AzureSqlDatabaseBackupLongTermRetentionPolicyModel>();
+            results.Add(ModelAdapter.GetDatabaseBackupLongTermRetentionPolicy(this.ResourceGroupName, this.ServerName, this.DatabaseName));
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureSqlServerBackupLongTermRetentionVault.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+using System.Linq;
+using System.Collections.Generic;
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Sql.Backup.Model;
+using Microsoft.Azure.Commands.Sql.Database.Model;
+
+namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
+{
+    [Cmdlet(VerbsCommon.Get, "AzureRmSqlServerBackupLongTermRetentionVault",
+         ConfirmImpact = ConfirmImpact.None)]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/GetAzureSqlServerBackupLongTermRetentionVault.cs]: Remove ConfirmImpact change

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+using Microsoft.Azure.Commands.Sql.Backup.Model;
+using Microsoft.Azure.Commands.Sql.Database.Model;
+
+namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
+{
+    /// <summary>
+    /// Cmdlet to create or update a new Azure Sql Database backup archival policy
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "AzureRmSqlDatabaseBackupLongTermRetentionPolicy",
+        ConfirmImpact = ConfirmImpact.Medium)]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        /// <summary>
+        /// Create the model from user input
+        /// </summary>
+        /// <param name="model">Model retrieved from service</param>
+        /// <returns>The model that was passed in</returns>
+        protected override IEnumerable<AzureSqlDatabaseBackupLongTermRetentionPolicyModel> ApplyUserInputToModel(IEnumerable<AzureSqlDatabaseBackupLongTermRetentionPolicyModel> model)
+        {
+            List<Model.AzureSqlDatabaseBackupLongTermRetentionPolicyModel> newEntity = new List<AzureSqlDatabaseBackupLongTermRetentionPolicyModel>();
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlServerBackupLongTermRetentionVault.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+using Microsoft.Azure.Commands.Sql.Backup.Model;
+using Microsoft.Azure.Commands.Sql.Database.Model;
+
+namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
+{
+    /// <summary>
+    /// Cmdlet to create or update a new Azure Sql Server backup archival vault
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "AzureRmSqlServerBackupLongTermRetentionVault",
+        ConfirmImpact = ConfirmImpact.Medium)]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlServerBackupLongTermRetentionVault.cs]: Same comments as above

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Services/AzureSqlDatabaseBackupAdapter.cs [hunk 13]
```diff
@@ -176,7 +176,100 @@ internal AzureSqlDeletedDatabaseBackupModel GetDeletedDatabaseBackup(string reso
...
+            };
+        }
+
+        /// <summary>
+        /// Get a backup LongTermRetention vault for a given Azure SQL Server
+        /// </summary>
+        /// <param name="resourceGroup">The name of the resource group</param>
+        /// <param name="serverName">The name of the Azure SQL Server</param>
+        /// <returns>A backup vault</returns>
+        internal AzureSqlServerBackupLongTermRetentionVaultModel GetBackupLongTermRetentionVault(string resourceGroup, string serverName)
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Services/AzureSqlDatabaseBackupAdapter.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Services/AzureSqlDatabaseBackupCommunicator.cs [hunk 14]
```diff
@@ -117,6 +117,45 @@ public Management.Sql.Models.DeletedDatabaseBackup GetDeletedDatabaseBackup(stri
         }
 
         /// <summary>
+        /// Get a backup LongTermRetention vault for a given Azure SQL Server
+        /// </summary>
+        /// <param name="resourceGroup">The name of the resource group</param>
+        /// <param name="serverName">The name of the Azure SQL Server</param>
+        /// <returns>A backup vault</returns>
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Services/AzureSqlDatabaseBackupCommunicator.cs]: line breaks

**Comment by markcowl** [comment]
@hchungmsft some comments, please address

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Sql.Backup.Model;
+using Microsoft.Azure.Commands.Sql.Database.Model;
+
+namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
+{
+    /// <summary>
+    /// Cmdlet to create or update a new Azure Sql Database backup archival policy
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "AzureRmSqlDatabaseBackupLongTermRetentionPolicy")]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlDatabaseBackupLongTermRetentionPolicy.cs]: SupportsShouldProcess=true.  Need to use SHouldProcess to implement Confirm/WhatIf before cmdlet execution

File: src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlServerBackupLongTermRetentionVault.cs [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Sql.Backup.Model;
+using Microsoft.Azure.Commands.Sql.Database.Model;
+
+namespace Microsoft.Azure.Commands.Sql.Backup.Cmdlet
+{
+    /// <summary>
+    /// Cmdlet to create or update a new Azure Sql Server backup archival vault
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "AzureRmSqlServerBackupLongTermRetentionVault")]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/Database Backup/Cmdlet/SetAzureSqlServerBackupLongTermRetentionVault.cs]: Same

**Comment by markcowl** [comment]
@hchungmsft The only rermaining issue is supporting ShouldProcess for set cmdlets - this was in my previous comments as well.  Please address this and this PR looks good to merge.

**Pull request merged**